_posts/2017-10-31-migrating-foreman-from-mysql-to-postgresql.md

```diff
@@ -0,0 +1,133 @@
+---
+layout: post
+title: Migrating Foreman from MySQL to Postgresql
+date: 2017-10-31 11:55:05
+author: Timo Goebel
+tags:
+- foreman
+---
+
+While Foreman supports both MySQL and Postgresql as a database management system, Katello requires Postgresql. If you want to add the Katello plugin on an existing Foreman setup with MySQL, you need to migrate your database. But there are other reasons why you might want to switch your database management system. This blog post describes what steps are required to migrate your data.
+
+
+While this blog post describes the required steps to migrate from MySQL to Postgresql, it can easily be adapted to migrate from Postgresql to MySQL.
+
+## Be careful
+
+First of all, a word of caution: This is not for the faint-hearted. And even fearless sysadmins should take backups before starting with this task. There is no guarantee, this will work. These steps have been tested with Foreman 1.15 but should work with other versions as well.
+
+## Motivation
+
+Postgresql and MySQL are both battle-tested open source database management systems. They have specific advantages and disadvantages. The most important reason to prefer one over the other is the available know-how in your organization. If you have a dedicated MySQL team that can provide a database for you, then you should by all means prefer that.
+If you do not have anybody who can provide you with a database on either MySQL or Postgresql, I'd tend to the latter as Postgresql seems a little more robust.
+
+## Prerequisites
+
+This blog post assumes that you already have the source and target databases set up. Make sure to setup the Postgresql database with UTF-8 enabled.
+
+First of all, you need to install support for either Postgresql or MySQL to Foreman. One of them will be missing. `yum install -y foreman-postgresql` adds support for Postgresql, `yum install -y foreman-mysql2` does the same for MySQL.
+
+Next, stop all your foreman instances and make sure, they stay stopped for good. On RedHat systems, you need to stop the `httpd` and (if applicable) `foreman-tasks` service. Make sure that neither Puppet or a monitoring system automatically starts these services again.
+
+You can issue the MySQL command `SHOW FULL PROCESSLIST;` to show you all active connections to your MySQL database. Make sure, there are none.
+
+If you are using the foreman-tasks plugin, you need to patch the script we are using to migrate the data. If you are not sure, wether you use this plugin, check if you have an entry `Tasks` in Foreman's Monitor menu. If you are still not sure, it is safe to apply the patch in any case.
+
+```sh
+wget -O /tmp/convert-tasks-patch.diff https://gist.githubusercontent.com/timogoebel/cd29e5206678ca0e0d78222af70583ec/raw/bbe998624098c1c3e656fe2582d33b0d9f9285aa/0001-convert-skip-foreman-tasks-tables.patch
```

Comment:
Is this patch upstream included? Should it be noted that it's only for 1.15?

Reply:
It's not included upstream. We should provide a way for plugins to specify tables that should be skipped. But that's not trivial to do.
dynflow defines tables that are outside of active record's control. As dynflow is now part of core, I believe we could skip all dynflow tables there. That would be a simple change.